tests/unit/lms/services/jwt_test.py

```diff
@@ -107,3 +109,20 @@ def test_it_uses_custom_lifetime(self):
     def _jwt_lifetime(payload):
         now = datetime.datetime.now(datetime.timezone.utc).timestamp()
         return payload["exp"] - now
+
+
+@pytest.fixture
+def svc():
+    return JWTService()
+
+
+class TestFactory:
+    def test_it(self, pyramid_request, JWTService):
+        jwt_service = factory(sentinel.context, pyramid_request)
+
+        JWTService.assert_called_once()
+        assert jwt_service == JWTService.return_value
+
+    @pytest.fixture
+    def JWTService(self, patch):
+        return patch("lms.services.jwt.JWTService")
```

Comment:
This all seems like a lot of test for not very much code. If you count out the comments there's not very much going on in that file. I've had a go at rewriting the first set of tests to trim down on the volume.

Some rules of thumbs I've noticed:

 * You ideally want your tests to be ~2x the length of the code
 * Above 3x gets really hard to maintain
 * For example, in this code it's hard to spot that we have the same test twice (take a look!)

Here's my attempt:

```python
class TestDecodeWithSecret:
    def test_it_can_decode_a_jwt_from_encode_jwt(self, svc):
        original_payload = {"key": "value"}
        jwt_str = svc.encode_with_secret(original_payload, "test_secret")

        decoded_payload = svc.decode_with_secret(jwt_str, "test_secret")

        assert decoded_payload == original_payload

    @pytest.mark.parametrize(
        "error,kwargs",
        (
            param(InvalidJWTError, {"secret": "WRONG"}, id="wrong_secret"),
            param(InvalidJWTError, {"algorithm": "HS384"}, id="wrong_algorithm"),
            param(InvalidJWTError, {"lifetime": None}, id="no_expiry"),
            param(
                ExpiredJWTError,
                {"lifetime": timedelta(hours=-1)},
                id="already_expired",
            ),
        ),
    )
    def test_it_raises_for_bad_jwts(self, svc, error, kwargs):
        jwt_str = self.encode_jwt({"key": "value"}, **kwargs)

        with pytest.raises(error):
            svc.decode_with_secret(jwt_str, "test_secret")

    def encode_jwt(
        self,
        payload,
        lifetime=timedelta(hours=1),
        secret="test_secret",
        algorithm="HS256",
    ):
        """Return payload encoded to a jwt with secret and algorithm."""
        payload = copy.deepcopy(payload)

        if lifetime:
            payload["exp"] = datetime.utcnow() + lifetime

        return jwt.encode(payload, secret, algorithm=algorithm)
```

We can also see the bigness is begetting more bigness. As the tests had run so long, they were split in two which (very slightly) increases the size again. If we can get the length under-control they could probably be one test class with `test_decode_with_secret_*` instead of `test_it_` again.

The above is a very straight refactor of the existing tests, but there's also an element to which we seem to be testing the underlying `jwt` library more than our own code. I think if some of the tests used a patched `jwt` library they could be cut down further to a few intergrated tests (like these) and probably one `assert_called_once_with` etc. You'd have to suck it an see though; it could make it worse.